File: cloud-functions/lib/tfidf.py
import math
import re
from collections import Counter
from typing import List, Dict

def tokenize(text: str) -> List[str]:
    return re.findall(r'\w+', text.lower())

def compute_tf(text: str) -> Dict[str, float]:
    tokens = tokenize(text)
    total = len(tokens)
    if total == 0:
        return {}
    counts = Counter(tokens)
    return {word: count / total for word, count in counts.items()}

def compute_idf(documents: List[str]) -> Dict[str, float]:
    n = len(documents)
    if n == 0:
        return {}
    doc_freq: Dict[str, int] = {}
    for doc in documents:
        words = set(tokenize(doc))
        for word in words:
            doc_freq[word] = doc_freq.get(word, 0) + 1
    return {word: math.log(n / (1 + df)) + 1 for word, df in doc_freq.items()}

def compute_tfidf(text: str, idf: Dict[str, float]) -> Dict[str, float]:
    tf = compute_tf(text)
    return {word: tf_val * idf.get(word, 0.01) for word, tf_val in tf.items()}

def cosine_similarity(vec1: Dict[str, float], vec2: Dict[str, float]) -> float:
    all_words = set(vec1.keys()) | set(vec2.keys())
    dot = sum(vec1.get(w, 0) * vec2.get(w, 0) for w in all_words)
    norm1 = math.sqrt(sum(v * v for v in vec1.values()))
    norm2 = math.sqrt(sum(v * v for v in vec2.values()))
    if norm1 == 0 or norm2 == 0:
        return 0.0
    return dot / (norm1 * norm2)

def jaccard_similarity(set1: set, set2: set) -> float:
    if not set1 and not set2:
        return 0.0
    union = set1 | set2
    inter = set1 & set2
    if not union:
        return 0.0
    return len(inter) / len(union)

def skor_pendidikan(pendidikan: str, minimal: str) -> float:
    TINGKAT = ["SD", "SMP", "SMA/SMK", "D1/D3", "S1/D4", "S2", "S3"]
    try:
        idx = TINGKAT.index(pendidikan)
        min_idx = TINGKAT.index(minimal)
    except ValueError:
        return 0.0
    if idx >= min_idx:
        return 100.0
    return max(0.0, 100.0 - (min_idx - idx) * 25.0)

def skor_umur(umur: int, min_u: int, max_u: int) -> float:
    if min_u <= umur <= max_u:
        return 100.0
    if umur < min_u:
        return max(0.0, 100.0 - (min_u - umur) * 10.0)
    return max(0.0, 100.0 - (umur - max_u) * 10.0)

File: src/lib/tfidf.ts
export function computeTf(text: string): Record<string, number> {
  const tokens = text.toLowerCase().match(/\w+/g) || []
  if (!tokens.length) return {}
  const cnt: Record<string, number> = {}
  for (const t of tokens) cnt[t] = (cnt[t] || 0) + 1
  const n = tokens.length
  const res: Record<string, number> = {}
  for (const [k, v] of Object.entries(cnt)) res[k] = v / n
  return res
}

export function cosineSimilarity(
  a: Record<string, number>,
  b: Record<string, number>
): number {
  const keys = new Set([...Object.keys(a), ...Object.keys(b)])
  let dot = 0, na = 0, nb = 0
  for (const k of keys) {
    const av = a[k] || 0
    const bv = b[k] || 0
    dot += av * bv
    na += av * av
    nb += bv * bv
  }
  na = Math.sqrt(na)
  nb = Math.sqrt(nb)
  return na && nb ? dot / (na * nb) : 0
}

export function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  let inter = 0
  for (const x of a) if (b.has(x)) inter++
  const uni = a.size + b.size - inter
  return uni ? inter / uni : 0
}

export const TINGKAT_PENDIDIKAN = [
  "SD", "SMP", "SMA/SMK", "D1/D3", "S1/D4", "S2", "S3",
]

export function skorPendidikan(pendidikan: string, minimal: string): number {
  const idx = TINGKAT_PENDIDIKAN.indexOf(pendidikan)
  const minIdx = TINGKAT_PENDIDIKAN.indexOf(minimal)
  if (idx < 0 || minIdx < 0) return 0
  if (idx >= minIdx) return 100
  return Math.max(0, 100 - (minIdx - idx) * 25)
}

export function skorUmur(umur: number, min: number, max: number): number {
  if (umur >= min && umur <= max) return 100
  if (umur < min) return Math.max(0, 100 - (min - umur) * 10)
  return Math.max(0, 100 - (umur - max) * 10)
}
